### PR TITLE
feat: agent in-flight memory writes (Provisional notes, recency-wins injection, MM hardening)

### PR DESCRIPTION
**Pull request by AN-BC**
## Summary

Implements the **agent in-flight memory writes** capability planned in this PR (plan doc updated in-place to the reviewed design). Agents with `AllowMemoryWrite=1` can now commit durable, cross-run memories mid-run via a `memoryWrites` field on the loop response — landing as **`Status='Provisional'` notes that are immediately injectable** and later hardened or pruned by the Memory Manager.

Implements the design **as amended in this PR's review discussion**:
- `Status='Provisional'` lifecycle state (not a `HardenedAt` timestamp column)
- Provisional notes inject with **higher precedence — recency wins** (a just-stated preference beats stale vetted memory); MM treats them as suggestions to reconcile
- Same-run conflicts use **supersede-own** (last write wins within a run); cross-run conflicts defer to MM
- `ExpiresAt` cleared to NULL on hardening (decay-managed like all Active notes)

## What's in it

**Schema** (`V202606111936__v5.41.x`): widen `AIAgentNote.Status` CHECK with `Provisional`; add `AIAgentNote.AuthoredBy` (`Agent|MemoryManager|User`, default `MemoryManager`); add `AIAgent.AllowMemoryWrite` (default 0). DDL + extended properties only; CodeGen output committed as the accompanying `Metadata_Sync` migration (current per-release convention), generated entity/resolver/form files included.

**Write path**: `MemoryWriteManager` (artifact-tool-manager pattern) with a framework-enforced guard pipeline — type restriction (`Preference`/`Context` only, prompt-injection defense), within-run idempotency, per-run cap (5), scope clamp (≤ Agent+User, hints can only narrow), near-dup vector guard (0.85) with supersede-own and fail-open on vector outage, full provenance stamping + 7-day TTL. BaseAgent processes writes inline (zero turn cost) via `processMemoryWritesForTurn`, records each as a `Tool` run step, and reports results back to the agent via an expiring message. Writes run sequentially by design (each save syncs the vector store the next write's dedupe check reads — commented in code). Multi-provider safe: persistence goes through the caller's provider (`context.provider ?? new Metadata()`), with BaseAgent passing its `ProviderToUse`.

**Read path**: shared `INJECTABLE_NOTE_STATUSES` (`Active`,`Provisional`) in `@memberjunction/core-entities` drives every filter — vector-service sync gate, embeddings refresh, fallback cache, SQL scoping, MM TTL pruning. `FormatNotesForInjection` renders provisional notes first with `(provisional)` labels and the amended recency-wins `<memory_policy>`.

**Hardening**: new MM pass runs unconditionally at the start of every cycle — LLM-dedupes provisional notes against hardened ones (duplicates archived with lineage), hardens survivors to `Active`/`ExpiresAt=NULL` so they join importance/consolidation/contradiction/decay in the same cycle.

**Prompt template**: gated `memoryWrites` response field + "Durable Memory" docs section — only rendered for opted-in agents (`AllowMemoryWrite` gates the template variable AND the turn-loop execution; disabled agents never see the capability and drift is rejected with one observable skip step).

**Docs**: `packages/AI/Agents/readme.md` gains MemoryWriteManager + hardening-pass sections; plan doc updated to the as-built design.

## Test Plan

**Automated (Claude-run, locally against a clean v5.41 database):**
- [x] Unit suites green across all four modified packages — ai-agents **1162**, core-entities **463**, core-entities-server **199**, ai-engine **75** (29 new tests: guard pipeline, threading through all four next-step shapes, provisional-first formatting)
- [x] `MJGlobal` MultiProviderCompliance scanner passes (provider threaded through writes)
- [x] E2E smoke harness `scripts/memory-write-smoke.ts` (full server pipeline, live DB): **21/21** — provisional persistence w/ embedding + scope + TTL, within-run idempotency, supersede-own, cross-run dedupe, provisional-first injection, hardening flip to Active
- [x] Migration applied to a clean database; schema constraints + EntityFieldValue rows verified via sqlcmd
- [x] Re-merged latest `next` and re-ran builds + suites

**Live verification (Claude-driven browser via Playwright, watched by @arie-glazier-bcc):**
- [x] MJExplorer + MJAPI against a sandboxed DB; Sage (`AllowMemoryWrite=1`) extracted two memories from one natural-language message — verified as `Provisional`/`AuthoredBy='Agent'` rows + `Memory Write` run steps in the DB
- [x] Fresh conversation: both notes injected (verified verbatim in the persisted `AIPromptRun.Messages` payload — `📝 RECENT NOTES (2 provisional)` block + recency-wins policy) and Sage applied the preferences unprompted

## Notes for reviewers

- **Build engineer**: this PR adds one SQL Server migration (`V202606111936__v5.41.x__Agent_InFlight_Memory_Writes.sql`, schema DDL with the CodeGen output appended) — flagging for the PG migration process.
- **Trust-posture note (deliberate)**: recency-wins means an unvetted provisional note briefly outranks vetted notes. Contained by: descriptive-types-only guard, scope clamp (never global/company-wide), per-run cap, 7-day TTL self-cleaning, and MM review within one cycle. This inverts the original plan's lower-precedence posture per the review discussion.
- **Found during live testing, not in this PR**: `RunAIAgentResolver.loadConversationHistoryWithAttachments` includes the conversations UI's `In-Progress` placeholder row in agent history, so the message array ends with an assistant turn — Anthropic 4.6-era models hard-reject this (`400: This model does not support assistant message prefill`). One-line fix validated locally; held for a separate PR. Related: `AIModelVendor/AIModel.SupportsPrefill` metadata is stale for current Anthropic models.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Post-review hardening (from live UI edge-case testing)

A structured edge-case campaign (cross-conversation provenance forensics, contradiction sequences, cap overflow, prompt-injection attempts, revoke kill-tests, gate-off probes) confirmed the core loop and surfaced four contours; three are fixed in this PR, one is tracked:

- **Dedupe narrowed to exact restatements (ADD-only-strict)** — similarity-only dedupe absorbed corrections into the stale opposite note ("never use pie charts" reinforced "loves pie charts", which the scheduled MM then hardened). Cross-run `deduped` now requires exact normalized-text equality; any textual difference writes a new provisional note, and a new smoke step verifies the Memory Manager's hardening dedupe consolidates the paraphrases this admits (within-batch vector-cache visibility confirmed).
- **Dates rendered in note injection** (`[Preference, 2026-06-10] …`) — contradictory Active notes can coexist between the every-cycle hardening pass and the daily contradiction phase; the policy's most-recent-wins tiebreaker is now resolvable. Absolute dates keep the injected message byte-stable between note-set changes.
- **No-premature-save prompt rule** — agents claimed "I've saved all seven facts" when two were `skipped-cap` (replies are composed before disposition feedback arrives next turn). The Durable Memory docs now forbid claiming a save before the result message.
- **Gate-off agents hallucinate saving** ("Got it! I've saved your preference" with no tool present, nothing persisted) — tracked as #2833; always-on prompt hint rejected for token cost.

Verification: ai-agents suite 1167 passing (+5 new); smoke harness 26/26 (+5 new checks incl. the paraphrase-written regression and within-batch hardening dedupe); UI re-test of the original pie/bar correction scenario.

### CODE REVIEW

**Comment by AN-BC** [comment]
@arie-glazier-bcc please share your thoughts on this, pretty simple PR but need to make sure this ties into the overall memory architecture well. CC thinks so. I tend to agree. Your thoughts?

**Comment by arie-glazier-bcc** [comment]
I think this makes sense overall.  A couple questions:

* Lower precedence - Would we want this to actually be higher precedence?  If a memory is so important that it needs to be written and used *right now*, then I would think it should be at the top?  e.g. using your example, if we have a history of memories saying "we like red charts" but then in the conversation we say "actually, i discovered i really love blue charts" then i would think we want that at the top?

* I have a bit of a nitpick around `HardenedAt`.  I don't quite get what this buys us over an additional `Status` value of `Provisional`.

* I think we need to think about / be careful about edge cases and multiple direct memory writes.  What happens when we say "I love blue charts!" and then "Oh jeez, thats actually very ugly.  I love red charts!"?  This makes me think we might want to allow agents to pass off *potential* memories to another agent and not have each agent make that decision itself?

**Comment by AN-BC** [comment]
> I think this makes sense overall. A couple questions:
> 
> * Lower precedence - Would we want this to actually be higher precedence?  If a memory is so important that it needs to be written and used _right now_, then I would think it should be at the top?  e.g. using your example, if we have a history of memories saying "we like red charts" but then in the conversation we say "actually, i discovered i really love blue charts" then i would think we want that at the top?
> * I have a bit of a nitpick around `HardenedAt`.  I don't quite get what this buys us over an additional `Status` value of `Provisional`.
> * I think we need to think about / be careful about edge cases and multiple direct memory writes.  What happens when we say "I love blue charts!" and then "Oh jeez, thats actually very ugly.  I love red charts!"?  This makes me think we might want to allow agents to pass off _potential_ memories to another agent and not have each agent make that decision itself?

- I think you might be right about higher precedence. In this case if another similar memory exists with lower, would it get merged **into** the newly formed memory? 
- I like `Status=='Provisional'` too
- the "potential" concept might be even more apt than "provisional" actually and the thing is the immediate writes to memory are really that. They're intermediate where they might actually end up being reasoned over by the memory manager as inconsistent, out of date, or in conflict and "lose" to other memories predating or postdating it. So i think "potential" or "provisional" memories do have effect immediately for new agent runs but should be considered by the memory manager as **suggestions** as opposed to ground truth, which of course perhaps is contradicatory to the higehr precedence point??

File: migrations/v5/V202606111936__v5.41.x__Agent_InFlight_Memory_Writes.sql [hunk 1]
```diff
@@ -0,0 +1,8665 @@
...
+    AuthoredBy NVARCHAR(20) NOT NULL
+    CONSTRAINT DF_AIAgentNote_AuthoredBy DEFAULT 'MemoryManager'
+    CONSTRAINT CK_AIAgentNote_AuthoredBy CHECK (AuthoredBy IN ('Agent', 'MemoryManager', 'User'));
+GO
+
+-- 3. AllowMemoryWrite: per-agent opt-in gate for in-flight memory writes,
+--    following the InjectNotes gating pattern. Off by default.
+ALTER TABLE ${flyway:defaultSchema}.AIAgent ADD
+    AllowMemoryWrite BIT NOT NULL
+    CONSTRAINT DF_AIAgent_AllowMemoryWrite DEFAULT 0;
```
Comment (AN-BC) [inline migrations/v5/V202606111936__v5.41.x__Agent_InFlight_Memory_Writes.sql]: wonder if this should flip to default to true?

File: migrations/v5/V202606111936__v5.41.x__Agent_InFlight_Memory_Writes.sql [hunk 2]
```diff
@@ -0,0 +1,8665 @@
...
+    ADD CONSTRAINT CK_AIAgentNote_Status
+    CHECK (Status IN ('Active', 'Pending', 'Revoked', 'Archived', 'Provisional'));
+GO
+
+-- 2. AuthoredBy: explicit provenance for who created the note. SourceAIAgentRunID
+--    alone can't distinguish written-in-flight-by-agent from extracted-from-a-run-
+--    by-MemoryManager (both carry a run ID). Default 'MemoryManager' backfills all
+--    legacy rows correctly (every existing auto-generated note came from MM).
+ALTER TABLE ${flyway:defaultSchema}.AIAgentNote ADD
+    AuthoredBy NVARCHAR(20) NOT NULL
```
Comment (AN-BC) [inline migrations/v5/V202606111936__v5.41.x__Agent_InFlight_Memory_Writes.sql]: this is a the logged in user at the time or something else?

File: migrations/v5/V202606111936__v5.41.x__Agent_InFlight_Memory_Writes.sql [hunk 3]
```diff
@@ -0,0 +1,8665 @@
...
+    @name = N'MS_Description',
+    @value = N'Lifecycle status of the note. Pending = awaiting approval, Active = vetted and injectable, Provisional = written in-flight by an agent (immediately injectable, awaiting Memory Manager hardening to Active), Revoked = superseded/withdrawn, Archived = retired by consolidation or decay.',
+    @level0type = N'SCHEMA', @level0name = N'${flyway:defaultSchema}',
+    @level1type = N'TABLE',  @level1name = N'AIAgentNote',
+    @level2type = N'COLUMN', @level2name = N'Status';
+GO
+
+EXEC sp_addextendedproperty
+    @name = N'MS_Description',
+    @value = N'Provenance of the note: Agent = written in-flight during an agent run, MemoryManager = extracted/consolidated by the scheduled Memory Manager, User = manually created by a person.',
```
Comment (AN-BC) [inline migrations/v5/V202606111936__v5.41.x__Agent_InFlight_Memory_Writes.sql]: oh I see here so this is AuthorType really not AuthoredBy wouldn't that make more sense with a CHECK constraint?

File: packages/AI/Agents/README.md [hunk 4]
```diff
@@ -225,6 +225,22 @@ Optimizes agent startup by batch-loading all required metadata in parallel:
 - Sub-agent data
 - Prompt configurations
 
+### MemoryWriteManager (In-Flight Memory Writes)
```
Comment (AN-BC) [inline packages/AI/Agents/README.md:232]: Memory, overall, is such a key topic I wonder if this deserves a much more detailed borken out guide we link to from here?

File: packages/AI/Agents/scripts/memory-write-smoke.ts [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+import { setupSQLServerClient, SQLServerProviderConfigData, UserCache } from '@memberjunction/sqlserver-dataprovider';
+import { Metadata, RunView, UserInfo, LogStatus } from '@memberjunction/core';
+import { MJAIAgentNoteEntity } from '@memberjunction/core-entities';
+import { AIEngine } from '@memberjunction/aiengine';
+import { MemoryWriteManager, MemoryWriteContext } from '@memberjunction/ai-agents';
+import { AgentContextInjector } from '@memberjunction/ai-agents';
+import { MemoryManagerAgent } from '@memberjunction/ai-agents';
+import { AIPromptRunner } from '@memberjunction/ai-prompts';
+import { MJAIPromptEntityExtended } from '@memberjunction/ai-core-plus';
+
```
Comment (AN-BC) [inline packages/AI/Agents/scripts/memory-write-smoke.ts:36]: test scripts shouldn't be in root, should be under section of repo related to, so this would be under packages/AI/Agents

**Review by AN-BC — COMMENTED** [review]
overall looks good, thoughts on my feedback @arie-glazier-bcc

File: migrations/v5/V202606122000__v5.41.x__Agent_InFlight_Memory_Writes.sql [hunk 6]
```diff
@@ -0,0 +1,10868 @@
...
+    AuthorType NVARCHAR(20) NOT NULL
+    CONSTRAINT DF_AIAgentNote_AuthorType DEFAULT 'MemoryManager'
+    CONSTRAINT CK_AIAgentNote_AuthorType CHECK (AuthorType IN ('Agent', 'MemoryManager', 'User'));
+GO
+
+-- 3. AllowMemoryWrite: per-agent gate for in-flight memory writes, following
+--    the InjectNotes gating pattern. ON by default — the Memory Manager already
+--    extracts memories for every agent, so direct writes are consistent with
+--    that posture; opt OUT for restricted or experimental agents.
+ALTER TABLE ${flyway:defaultSchema}.AIAgent ADD
```
Comment (AN-BC) [inline migrations/v5/V202606122000__v5.41.x__Agent_InFlight_Memory_Writes.sql:43]: @arie-glazier-bcc I have a PR in progress that also touches this table, so we'll need to put the `R` script contents in your migration, update to today's date so your script runs after mine and we'll have you do a quick test once I merge you can pull next and reapply all migrations to a clean database and see if it all works, It should if you paste the contenst of the `R` script on top of your migration - this forces metadata refresh before your script runs

File: .changeset/proud-otters-remember.md [hunk 7]
```diff
@@ -0,0 +1,12 @@
+---
+"@memberjunction/core": minor
```
Comment (AN-BC) [inline .changeset/proud-otters-remember.md:2]: I don't see any code changes to this package, you should remove it so it refelcts what was actually changed, and update below ones to minor

**Review by AN-BC — COMMENTED** [review]
LGTM overall @arie-glazier-bcc but wait to merge per below, I will be done with my other PR in flight today and you can pull `next` build a CLEAN db from scratch and run `mj migrate --dir ./metadata` and retest. Should be ok if you follow below

File: migrations/v5/V202606122000__v5.41.x__Agent_InFlight_Memory_Writes.sql (hunk 6 above)
Comment (AN-BC) [reply migrations/v5/V202606122000__v5.41.x__Agent_InFlight_Memory_Writes.sql:43]: actually this is incorrect, I didn't mutate that table so we are good. Merging this for you now @arie-glazier-bcc so we can get into 5.41

**Review by AN-BC — COMMENTED** [review]
see latest comment @arie-glazier-bcc
